enhancements/microshift/loadbalancer-service-support.md

```diff
@@ -0,0 +1,199 @@
+workloads/deployment for MicroShift as they use other k8s flavors as well. It is
+also promised by us that users do not have to modify workload manifests that
+work on OpenShift to run on MicroShift (as long as they do not use APIs
+unsupported by MicroShift).
+
+## Motivation
+
+Normally, the loadbalancer type Service is supported by a Kubernetes cluster
+with external load balancers. In Microshift, as there is only one node in the
+cluster, we don't really need to have a load balancer. However, from user
+experience perspective, it's important for customers to be able to migrate
+the workload to Microshift without modification. So we need to find a way to
+support the loadbalancer type Service without a load balancer.
+
+### User Stories
+
+As a Microshift user, I want to reuse the Service manifest that with type of
+`LoadBalancer` without modification. I don't want to allocate a new IP address
+to the Node as the loadbalancer VIP. I want to use the NodeIP as the ingress IP
+of the loadbalancer Service. 
+
+### Goals
+
+* Allow users to deploy the load balancer type Service to Microshift.
+* The node IP shall be used as the ingress IP of the Service.
+* As we will support pluggable CNI in Microshift. The implementation shall be
+also compatible with CNIs other than OVN-Kubernetes.
+
+### Non-Goals
+
+* 
+
+## Proposal
+
+To support the loadbalancer type Service, we need to create a new controller
+component that watch the Service objects. It shall be responsible for:
+
+1. work with CNIs to plumb the ingress traffic to the endpoint pods.
+2. publish the ingress endpoint information in the Service's
+   `.status.loadBalancer` field.
+
+### Workflow Description
+
+**cluster user** is a human user responsible for provisioning Services to a
+Microshift cluster.
+
+1. The cluster user deploys a Service of type: LoadBalancer 
+2. The loadbalancer service controller watches the creation of the service. It
+will update the Service's `.status.loadBalancer` field, and help the CNIs to 
+plumbing the ingress traffic to the endpoint pods.
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+This new controller is part of the microshift binary. It shall be able to know
+what CNI is used by the cluster, and behave differently based on the CNI type.
+
+### Risks and Mitigations
+
+The pluggable CNI design of Microshift is still under discussion, we need to
+make sure that CNI type can be aware in Microshift binary.
+
+### Drawbacks
+
+It is not a CNI agnostic design. The loadbalancer Service controller will behave
+differently between OVN-Kubernetes and other CNIs.
+
+## Design Details
+
+The Service of loadblancer type is designed to exposes the Service externally
+using a cloud provider's load balancer. Normally, there is a loadbalancer
+Service controller which is responsible for:
+
+1. provision a loadbalancer for the Service.
+2. publish the ingress endpoint information in the Service's
+   `.status.loadBalancer` field.
+
```

Comment:
Does the load balancer controller need to handle service port mapping?

Reply:
No, the loadbalancer itself would do that.